// @circle-fin/bridge-kit/chains.cjs
/**
 * Copyright (c) 2026, Circle Internet Group, Inc. All rights reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

var zod = require('zod');

/**
 * @packageDocumentation
 * @module ChainDefinitions
 *
 * This module provides a complete type system for blockchain chain definitions.
 * It supports both EVM and non‑EVM chains, token configurations, and multiple
 * versions of the Cross-Chain Transfer Protocol (CCTP). Additionally, utility types
 * are provided to extract subsets of chains (e.g. chains supporting USDC, EURC, or specific
 * CCTP versions) from a provided collection.
 *
 * All types are fully documented with TSDoc to maximize developer experience.
 */ // -----------------------------------------------------------------------------
// Currency and Base Types
// -----------------------------------------------------------------------------
/**
 * Represents basic information about a currency or token.
 * @category Types
 * @description Provides the essential properties of a cryptocurrency or token.
 * @example
 * ```typescript
 * const ethCurrency: Currency = {
 *   name: "Ether",
 *   symbol: "ETH",
 *   decimals: 18
 * };
 * ```
 */ var Blockchain;
(function(Blockchain) {
    Blockchain["Algorand"] = "Algorand";
    Blockchain["Algorand_Testnet"] = "Algorand_Testnet";
    Blockchain["Aptos"] = "Aptos";
    Blockchain["Aptos_Testnet"] = "Aptos_Testnet";
    Blockchain["Arc_Testnet"] = "Arc_Testnet";
    Blockchain["Arbitrum"] = "Arbitrum";
    Blockchain["Arbitrum_Sepolia"] = "Arbitrum_Sepolia";
    Blockchain["Avalanche"] = "Avalanche";
    Blockchain["Avalanche_Fuji"] = "Avalanche_Fuji";
    Blockchain["Base"] = "Base";
    Blockchain["Base_Sepolia"] = "Base_Sepolia";
    Blockchain["Celo"] = "Celo";
    Blockchain["Celo_Alfajores_Testnet"] = "Celo_Alfajores_Testnet";
    Blockchain["Codex"] = "Codex";
    Blockchain["Codex_Testnet"] = "Codex_Testnet";
    Blockchain["Cronos"] = "Cronos";
    Blockchain["Cronos_Testnet"] = "Cronos_Testnet";
    Blockchain["Edge"] = "Edge";
    Blockchain["Edge_Testnet"] = "Edge_Testnet";
    Blockchain["Ethereum"] = "Ethereum";
    Blockchain["Ethereum_Sepolia"] = "Ethereum_Sepolia";
    Blockchain["Hedera"] = "Hedera";
    Blockchain["Hedera_Testnet"] = "Hedera_Testnet";
    Blockchain["HyperEVM"] = "HyperEVM";
    Blockchain["HyperEVM_Testnet"] = "HyperEVM_Testnet";
    Blockchain["Injective"] = "Injective";
    Blockchain["Injective_Testnet"] = "Injective_Testnet";
    Blockchain["Ink"] = "Ink";
    Blockchain["Ink_Testnet"] = "Ink_Testnet";
    Blockchain["Linea"] = "Linea";
    Blockchain["Linea_Sepolia"] = "Linea_Sepolia";
    Blockchain["Monad"] = "Monad";
    Blockchain["Monad_Testnet"] = "Monad_Testnet";
    Blockchain["Morph"] = "Morph";
    Blockchain["Morph_Testnet"] = "Morph_Testnet";
    Blockchain["NEAR"] = "NEAR";
    Blockchain["NEAR_Testnet"] = "NEAR_Testnet";
    Blockchain["Noble"] = "Noble";
    Blockchain["Noble_Testnet"] = "Noble_Testnet";
    Blockchain["Optimism"] = "Optimism";
    Blockchain["Optimism_Sepolia"] = "Optimism_Sepolia";
    Blockchain["Pharos"] = "Pharos";
    Blockchain["Pharos_Testnet"] = "Pharos_Testnet";
    Blockchain["Polkadot_Asset_Hub"] = "Polkadot_Asset_Hub";
    Blockchain["Polkadot_Westmint"] = "Polkadot_Westmint";
    Blockchain["Plume"] = "Plume";
    Blockchain["Plume_Testnet"] = "Plume_Testnet";
    Blockchain["Polygon"] = "Polygon";
    Blockchain["Polygon_Amoy_Testnet"] = "Polygon_Amoy_Testnet";
    Blockchain["Sei"] = "Sei";
    Blockchain["Sei_Testnet"] = "Sei_Testnet";
    Blockchain["Solana"] = "Solana";
    Blockchain["Solana_Devnet"] = "Solana_Devnet";
    Blockchain["Sonic"] = "Sonic";
    Blockchain["Sonic_Testnet"] = "Sonic_Testnet";
    Blockchain["Stellar"] = "Stellar";
    Blockchain["Stellar_Testnet"] = "Stellar_Testnet";
    Blockchain["Sui"] = "Sui";
    Blockchain["Sui_Testnet"] = "Sui_Testnet";
    Blockchain["Unichain"] = "Unichain";
    Blockchain["Unichain_Sepolia"] = "Unichain_Sepolia";
    Blockchain["World_Chain"] = "World_Chain";
    Blockchain["World_Chain_Sepolia"] = "World_Chain_Sepolia";
    Blockchain["XDC"] = "XDC";
    Blockchain["XDC_Apothem"] = "XDC_Apothem";
    Blockchain["ZKSync_Era"] = "ZKSync_Era";
    Blockchain["ZKSync_Sepolia"] = "ZKSync_Sepolia";
})(Blockchain || (Blockchain = {}));
var SwapChain;
(function(SwapChain) {
    // Original 4 chains
    SwapChain["Ethereum"] = "Ethereum";
    SwapChain["Base"] = "Base";
    SwapChain["Polygon"] = "Polygon";
    SwapChain["Solana"] = "Solana";
    // Additional supported chains
    SwapChain["Arbitrum"] = "Arbitrum";
    SwapChain["Optimism"] = "Optimism";
    SwapChain["Avalanche"] = "Avalanche";
    SwapChain["Linea"] = "Linea";
    SwapChain["Ink"] = "Ink";
    SwapChain["World_Chain"] = "World_Chain";
    SwapChain["Unichain"] = "Unichain";
    SwapChain["Plume"] = "Plume";
    SwapChain["Sei"] = "Sei";
    SwapChain["Sonic"] = "Sonic";
    SwapChain["XDC"] = "XDC";
    SwapChain["HyperEVM"] = "HyperEVM";
    SwapChain["Monad"] = "Monad";
    // Testnet chains with swap support
    SwapChain["Arc_Testnet"] = "Arc_Testnet";
})(SwapChain || (SwapChain = {}));
exports.BridgeChain = void 0;
(function(BridgeChain) {
    // Mainnet chains with CCTPv2 support
    BridgeChain["Arbitrum"] = "Arbitrum";
    BridgeChain["Avalanche"] = "Avalanche";
    BridgeChain["Base"] = "Base";
    BridgeChain["Codex"] = "Codex";
    BridgeChain["Cronos"] = "Cronos";
    BridgeChain["Edge"] = "Edge";
    BridgeChain["Ethereum"] = "Ethereum";
    BridgeChain["HyperEVM"] = "HyperEVM";
    BridgeChain["Injective"] = "Injective";
    BridgeChain["Ink"] = "Ink";
    BridgeChain["Linea"] = "Linea";
    BridgeChain["Monad"] = "Monad";
    BridgeChain["Morph"] = "Morph";
    BridgeChain["Optimism"] = "Optimism";
    BridgeChain["Pharos"] = "Pharos";
    BridgeChain["Plume"] = "Plume";
    BridgeChain["Polygon"] = "Polygon";
    BridgeChain["Sei"] = "Sei";
    BridgeChain["Solana"] = "Solana";
    BridgeChain["Sonic"] = "Sonic";
    BridgeChain["Unichain"] = "Unichain";
    BridgeChain["World_Chain"] = "World_Chain";
    BridgeChain["XDC"] = "XDC";
    // Testnet chains with CCTPv2 support
    BridgeChain["Arc_Testnet"] = "Arc_Testnet";
    BridgeChain["Arbitrum_Sepolia"] = "Arbitrum_Sepolia";
    BridgeChain["Avalanche_Fuji"] = "Avalanche_Fuji";
    BridgeChain["Base_Sepolia"] = "Base_Sepolia";
    BridgeChain["Codex_Testnet"] = "Codex_Testnet";
    BridgeChain["Cronos_Testnet"] = "Cronos_Testnet";
    BridgeChain["Edge_Testnet"] = "Edge_Testnet";
    BridgeChain["Ethereum_Sepolia"] = "Ethereum_Sepolia";
    BridgeChain["HyperEVM_Testnet"] = "HyperEVM_Testnet";
    BridgeChain["Injective_Testnet"] = "Injective_Testnet";
    BridgeChain["Ink_Testnet"] = "Ink_Testnet";
    BridgeChain["Linea_Sepolia"] = "Linea_Sepolia";
    BridgeChain["Monad_Testnet"] = "Monad_Testnet";
    BridgeChain["Morph_Testnet"] = "Morph_Testnet";
    BridgeChain["Optimism_Sepolia"] = "Optimism_Sepolia";
    BridgeChain["Pharos_Testnet"] = "Pharos_Testnet";
    BridgeChain["Plume_Testnet"] = "Plume_Testnet";
    BridgeChain["Polygon_Amoy_Testnet"] = "Polygon_Amoy_Testnet";
    BridgeChain["Sei_Testnet"] = "Sei_Testnet";
    BridgeChain["Solana_Devnet"] = "Solana_Devnet";
    BridgeChain["Sonic_Testnet"] = "Sonic_Testnet";
    BridgeChain["Unichain_Sepolia"] = "Unichain_Sepolia";
    BridgeChain["World_Chain_Sepolia"] = "World_Chain_Sepolia";
    BridgeChain["XDC_Apothem"] = "XDC_Apothem";
})(exports.BridgeChain || (exports.BridgeChain = {}));
var UnifiedBalanceChain;
(function(UnifiedBalanceChain) {
    // Mainnet chains with Gateway V1 support
    UnifiedBalanceChain["Arbitrum"] = "Arbitrum";
    UnifiedBalanceChain["Avalanche"] = "Avalanche";
    UnifiedBalanceChain["Base"] = "Base";
    UnifiedBalanceChain["Ethereum"] = "Ethereum";
    UnifiedBalanceChain["HyperEVM"] = "HyperEVM";
    UnifiedBalanceChain["Optimism"] = "Optimism";
    UnifiedBalanceChain["Polygon"] = "Polygon";
    UnifiedBalanceChain["Sei"] = "Sei";
    UnifiedBalanceChain["Solana"] = "Solana";
    UnifiedBalanceChain["Sonic"] = "Sonic";
    UnifiedBalanceChain["Unichain"] = "Unichain";
    UnifiedBalanceChain["World_Chain"] = "World_Chain";
    // Testnet chains with Gateway V1 support
    UnifiedBalanceChain["Arbitrum_Sepolia"] = "Arbitrum_Sepolia";
    UnifiedBalanceChain["Arc_Testnet"] = "Arc_Testnet";
    UnifiedBalanceChain["Avalanche_Fuji"] = "Avalanche_Fuji";
    UnifiedBalanceChain["Base_Sepolia"] = "Base_Sepolia";
    UnifiedBalanceChain["Ethereum_Sepolia"] = "Ethereum_Sepolia";
    UnifiedBalanceChain["HyperEVM_Testnet"] = "HyperEVM_Testnet";
    UnifiedBalanceChain["Optimism_Sepolia"] = "Optimism_Sepolia";
    UnifiedBalanceChain["Polygon_Amoy_Testnet"] = "Polygon_Amoy_Testnet";
    UnifiedBalanceChain["Sei_Testnet"] = "Sei_Testnet";
    UnifiedBalanceChain["Solana_Devnet"] = "Solana_Devnet";
    UnifiedBalanceChain["Sonic_Testnet"] = "Sonic_Testnet";
    UnifiedBalanceChain["Unichain_Sepolia"] = "Unichain_Sepolia";
    UnifiedBalanceChain["World_Chain_Sepolia"] = "World_Chain_Sepolia";
})(UnifiedBalanceChain || (UnifiedBalanceChain = {}));
var EarnChain;
(function(EarnChain) {
    EarnChain["Arc_Testnet"] = "Arc_Testnet";
})(EarnChain || (EarnChain = {}));
/**
 * Blockchains supported as the source chain for cross-chain Earn deposits.
 *
 * Single source of truth for the source allowlist: the Earn Service bridge
 * route map is `satisfies`-checked against {@link EarnBridgeSourceBlockchain},
 * and both the type and the runtime validation schema derive from this array.
 *
 * @example
 * ```typescript
 * import { EARN_BRIDGE_SOURCE_BLOCKCHAINS } from '@core/chains'
 *
 * console.log(EARN_BRIDGE_SOURCE_BLOCKCHAINS.join(', '))
 * ```
 */ const EARN_BRIDGE_SOURCE_BLOCKCHAINS = [
    "Arbitrum_Sepolia",
    "Base_Sepolia",
    "Ethereum_Sepolia"
];
/**
 * Blockchains supported as the destination (vault) chain for cross-chain Earn
 * deposits. Single source of truth for the destination allowlist.
 *
 * @example
 * ```typescript
 * import { EARN_BRIDGE_DESTINATION_BLOCKCHAINS } from '@core/chains'
 *
 * console.log(EARN_BRIDGE_DESTINATION_BLOCKCHAINS.join(', '))
 * ```
 */ const EARN_BRIDGE_DESTINATION_BLOCKCHAINS = [
    "Arc_Testnet"
];

/**
 * Helper function to define a chain with proper TypeScript typing.
 *
 * This utility function works with TypeScript's `as const` assertion to create
 * strongly-typed, immutable chain definition objects. It preserves literal types
 * from the input and ensures the resulting object maintains all type information.
 *
 * When used with `as const`, it allows TypeScript to infer the most specific
 * possible types for all properties, including string literals and numeric literals,
 * rather than widening them to general types like string or number.
 * @typeParam T - The specific chain definition type (must extend ChainDefinition)
 * @param chain - The chain definition object, typically with an `as const` assertion
 * @returns The same chain definition with preserved literal types
 * @example
 * ```typescript
 * // Define an EVM chain with literal types preserved
 * const Ethereum = defineChain({
 *   type: 'evm',
 *   chain: Blockchain.Ethereum,
 *   chainId: 1,
 *   name: 'Ethereum',
 *   nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
 *   isTestnet: false,
 *   usdcAddress: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
 *   eurcAddress: null,
 *   cctp: {
 *     domain: 0,
 *     contracts: {
 *       TokenMessengerV1: '0xbd3fa81b58ba92a82136038b25adec7066af3155',
 *       MessageTransmitterV1: '0x0a992d191deec32afe36203ad87d7d289a738f81'
 *     }
 *   }
 * } as const);
 * ```
 */ function defineChain(chain) {
    return chain;
}

/**
 * Algorand Mainnet chain definition
 * @remarks
 * This represents the official production network for the Algorand blockchain.
 */ defineChain({
    type: 'algorand',
    chain: Blockchain.Algorand,
    name: 'Algorand',
    title: 'Algorand Mainnet',
    nativeCurrency: {
        name: 'Algo',
        symbol: 'ALGO',
        decimals: 6
    },
    isTestnet: false,
    explorerUrl: 'https://explorer.perawallet.app/tx/{hash}',
    rpcEndpoints: [
        'https://mainnet-api.algonode.cloud'
    ],
    eurcAddress: null,
    usdcAddress: '31566704',
    usdtAddress: null,
    cctp: null
});

/**
 * Algorand Testnet chain definition
 * @remarks
 * This represents the official testnet for the Algorand blockchain.
 */ defineChain({
    type: 'algorand',
    chain: Blockchain.Algorand_Testnet,
    name: 'Algorand Testnet',
    title: 'Algorand Test Network',
    nativeCurrency: {
        name: 'Algo',
        symbol: 'ALGO',
        decimals: 6
    },
    isTestnet: true,
    explorerUrl: 'https://testnet.explorer.perawallet.app/tx/{hash}',
    rpcEndpoints: [
        'https://testnet-api.algonode.cloud'
    ],
    eurcAddress: null,
    usdcAddress: '10458941',
    usdtAddress: null,
    cctp: null
});

/**
 * Aptos Mainnet chain definition
 * @remarks
 * This represents the official production network for the Aptos blockchain.
 */ defineChain({
    type: 'aptos',
    chain: Blockchain.Aptos,
    name: 'Aptos',
    title: 'Aptos Mainnet',
    nativeCurrency: {
        name: 'Aptos',
        symbol: 'APT',
        decimals: 8
    },
    isTestnet: false,
    explorerUrl: 'https://explorer.aptoslabs.com/txn/{hash}?network=mainnet',
    rpcEndpoints: [
        'https://fullnode.mainnet.aptoslabs.com/v1'
    ],
    eurcAddress: null,
    usdcAddress: '0xbae207659db88bea0cbead6da0ed00aac12edcdda169e591cd41c94180b46f3b',
    usdtAddress: '0x357b0b74bc833e95a115ad22604854d6b0fca151cecd94111770e5d6ffc9dc2b',
    cctp: {
        domain: 9,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9bce6734f7b63e835108e3bd8c36743d4709fe435f44791918801d0989640a9d',
                messageTransmitter: '0x177e17751820e4b4371873ca8c30279be63bdea63b88ed0f2239c2eea10f1772',
                confirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    }
});

/**
 * Aptos Testnet chain definition
 * @remarks
 * This represents the official test network for the Aptos blockchain.
 */ defineChain({
    type: 'aptos',
    chain: Blockchain.Aptos_Testnet,
    name: 'Aptos Testnet',
    title: 'Aptos Test Network',
    nativeCurrency: {
        name: 'Aptos',
        symbol: 'APT',
        decimals: 8
    },
    isTestnet: true,
    explorerUrl: 'https://explorer.aptoslabs.com/txn/{hash}?network=testnet',
    rpcEndpoints: [
        'https://fullnode.testnet.aptoslabs.com/v1'
    ],
    eurcAddress: null,
    usdcAddress: '0x69091fbab5f7d635ee7ac5098cf0c1efbe31d68fec0f2cd565e8d168daf52832',
    usdtAddress: null,
    cctp: {
        domain: 9,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x5f9b937419dda90aa06c1836b7847f65bbbe3f1217567758dc2488be31a477b9',
                messageTransmitter: '0x081e86cebf457a0c6004f35bd648a2794698f52e0dde09a48619dcd3d4cc23d9',
                confirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    }
});

/**
 * @packageDocumentation
 * @module SwapTokenRegistry
 *
 * Central swap token registry for swap-supported tokens.
 *
 * Contains metadata for all tokens that can be used in swap operations,
 * including decimals, categories, and descriptions. All packages should import
 * from this registry instead of maintaining separate token lists.
 */ /**
 * Swap token metadata.
 */ /**
 * Complete swap token registry - single source of truth for all swap-supported tokens.
 *
 * @remarks
 * All packages should import from this registry for swap operations.
 * Adding a new swap token requires updating only this registry.
 *
 * The NATIVE token is handled separately as it resolves dynamically based on chain.
 *
 * @example
 * ```typescript
 * import { SWAP_TOKEN_REGISTRY } from '@core/chains'
 *
 * // Get token decimals
 * const decimals = SWAP_TOKEN_REGISTRY.USDC.decimals  // 6
 *
 * // Check if token is stablecoin
 * const isStable = SWAP_TOKEN_REGISTRY.DAI.category === 'stablecoin'  // true
 * ```
 */ const SWAP_TOKEN_REGISTRY = {
    // ============================================================================
    // Stablecoins (6 decimals)
    // ============================================================================
    USDC: {
        symbol: 'USDC',
        decimals: 6,
        category: 'stablecoin',
        description: 'USD Coin'
    },
    EURC: {
        symbol: 'EURC',
        decimals: 6,
        category: 'stablecoin',
        description: 'Euro Coin'
    },
    USDT: {
        symbol: 'USDT',
        decimals: 6,
        category: 'stablecoin',
        description: 'Tether USD'
    },
    PYUSD: {
        symbol: 'PYUSD',
        decimals: 6,
        category: 'stablecoin',
        description: 'PayPal USD'
    },
    // ============================================================================
    // Stablecoins (18 decimals)
    // ============================================================================
    DAI: {
        symbol: 'DAI',
        decimals: 18,
        category: 'stablecoin',
        description: 'MakerDAO stablecoin'
    },
    USDE: {
        symbol: 'USDE',
        decimals: 18,
        category: 'stablecoin',
        description: 'Ethena USD (synthetic dollar)'
    },
    // ============================================================================
    // Wrapped Tokens
    // ============================================================================
    WBTC: {
        symbol: 'WBTC',
        decimals: 8,
        category: 'wrapped',
        description: 'Wrapped Bitcoin'
    },
    WETH: {
        symbol: 'WETH',
        decimals: 18,
        category: 'wrapped',
        description: 'Wrapped Ethereum'
    },
    WSOL: {
        symbol: 'WSOL',
        decimals: 9,
        category: 'wrapped',
        description: 'Wrapped Solana'
    },
    WAVAX: {
        symbol: 'WAVAX',
        decimals: 18,
        category: 'wrapped',
        description: 'Wrapped Avalanche'
    },
    WPOL: {
        symbol: 'WPOL',
        decimals: 18,
        category: 'wrapped',
        description: 'Wrapped Polygon'
    },
    CIRBTC: {
        symbol: 'CIRBTC',
        decimals: 8,
        category: 'wrapped',
        description: 'Circle Bitcoin'
    }
};
/**
 * Special NATIVE token constant for swap operations.
 *
 * @remarks
 * NATIVE is handled separately from SWAP_TOKEN_REGISTRY because it resolves
 * dynamically based on the chain (ETH on Ethereum, SOL on Solana, etc.).
 * Its decimals are chain-specific.
 */ const NATIVE_TOKEN = 'NATIVE';
/**
 * Array of all supported swap token symbols including NATIVE.
 * Useful for iteration, validation, and filtering.
 */ [
    ...Object.keys(SWAP_TOKEN_REGISTRY),
    NATIVE_TOKEN
];

/**
 * The bridge contract address for EVM testnet networks.
 *
 * This contract handles USDC transfers on testnet environments across
 * EVM-compatible chains. Use this address when deploying or testing
 * cross-chain USDC transfers on test networks.
 */ const BRIDGE_CONTRACT_EVM_TESTNET = '0xC5567a5E3370d4DBfB0540025078e283e36A363d';
/**
 * The bridge contract address for EVM mainnet networks.
 *
 * This contract handles USDC transfers on mainnet environments across
 * EVM-compatible chains. Use this address for production cross-chain
 * USDC transfers on live networks.
 */ const BRIDGE_CONTRACT_EVM_MAINNET = '0xB3FA262d0fB521cc93bE83d87b322b8A23DAf3F0';
/**
 * The adapter contract address for EVM mainnet networks.
 *
 * This contract serves as an adapter for integrating with various protocols
 * on EVM-compatible chains. Use this address for mainnet adapter integrations.
 */ const ADAPTER_CONTRACT_EVM_MAINNET = '0x7FB8c7260b63934d8da38aF902f87ae6e284a845';
/**
 * The adapter contract address for EVM testnet networks.
 *
 * This contract serves as an adapter for integrating with various protocols
 * on EVM-compatible testnet chains. Use this address for testnet adapter
 * integrations (e.g., Arc Testnet).
 */ const ADAPTER_CONTRACT_EVM_TESTNET = '0xBBD70b01a1CAbc96d5b7b129Ae1AAabdf50dd40b';
/**
 * The GatewayWallet contract address for EVM mainnet networks.
 *
 * This contract manages wallet operations for Gateway transactions
 * on mainnet environments across EVM-compatible chains.
 */ const GATEWAY_WALLET_EVM_MAINNET = '0x77777777Dcc4d5A8B6E418Fd04D8997ef11000eE';
/**
 * The GatewayMinter contract address for EVM mainnet networks.
 *
 * This contract handles minting operations for Gateway transactions
 * on mainnet environments across EVM-compatible chains.
 */ const GATEWAY_MINTER_EVM_MAINNET = '0x2222222d7164433c4C09B0b0D809a9b52C04C205';
/**
 * The GatewayWallet contract address for EVM testnet networks.
 *
 * This contract manages wallet operations for Gateway transactions
 * on testnet environments across EVM-compatible chains.
 */ const GATEWAY_WALLET_EVM_TESTNET = '0x0077777d7EBA4688BDeF3E311b846F25870A19B9';
/**
 * The GatewayMinter contract address for EVM testnet networks.
 *
 * This contract handles minting operations for Gateway transactions
 * on testnet environments across EVM-compatible chains.
 */ const GATEWAY_MINTER_EVM_TESTNET = '0x0022222ABE238Cc2C7Bb1f21003F0a260052475B';
/**
 * The GatewayWallet program address for Solana mainnet.
 *
 * This program manages wallet operations for Gateway transactions
 * on Solana mainnet.
 */ const GATEWAY_WALLET_SOLANA_MAINNET = 'GATEwy4YxeiEbRJLwB6dXgg7q61e6zBPrMzYj5h1pRXQ';
/**
 * The GatewayMinter program address for Solana mainnet.
 *
 * This program handles minting operations for Gateway transactions
 * on Solana mainnet.
 */ const GATEWAY_MINTER_SOLANA_MAINNET = 'GATEm5SoBJiSw1v2Pz1iPBgUYkXzCUJ27XSXhDfSyzVZ';
/**
 * The GatewayWallet program address for Solana devnet.
 *
 * This program manages wallet operations for Gateway transactions
 * on Solana devnet.
 */ const GATEWAY_WALLET_SOLANA_DEVNET = 'GATEwdfmYNELfp5wDmmR6noSr2vHnAfBPMm2PvCzX5vu';
/**
 * The GatewayMinter program address for Solana devnet.
 *
 * This program handles minting operations for Gateway transactions
 * on Solana devnet.
 */ const GATEWAY_MINTER_SOLANA_DEVNET = 'GATEmKK2ECL1brEngQZWCgMWPbvrEYqsV6u29dAaHavr';

/**
 * Arc Testnet chain definition
 * @remarks
 * This represents the test network for the Arc blockchain,
 * Circle's EVM-compatible Layer-1 designed for stablecoin finance
 * and asset tokenization. Arc uses USDC as the native gas token and
 * features the Malachite Byzantine Fault Tolerant (BFT) consensus
 * engine for sub-second finality.
 */ const ArcTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Arc_Testnet,
    name: 'Arc Testnet',
    title: 'ArcTestnet',
    nativeCurrency: {
        name: 'USDC',
        symbol: 'USDC',
        // Arc uses native USDC with 18 decimals for gas payments (EVM standard).
        // Note: The ERC-20 USDC contract at usdcAddress uses 6 decimals.
        // See: https://docs.arc.network/arc/references/contract-addresses
        decimals: 18
    },
    chainId: 5042002,
    isTestnet: true,
    explorerUrl: 'https://testnet.arcscan.app/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.testnet.arc.network/'
    ],
    eurcAddress: '0x89B50855Aa3bE2F677cD6303Cec089B5F319D72a',
    usdcAddress: '0x3600000000000000000000000000000000000000',
    usdtAddress: null,
    cctp: {
        domain: 26,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET,
        adapter: ADAPTER_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 26,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET,
                // DepositForHandler the GenericExecutor calls to run a fast cross-chain
                // deposit into the GatewayWallet above.
                depositForHandler: '0xD05E7D2E7d30b92c5F17d7d0fC575fce231F1A48'
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Arbitrum Mainnet chain definition
 * @remarks
 * This represents the official production network for the Arbitrum blockchain.
 */ const Arbitrum = defineChain({
    type: 'evm',
    chain: Blockchain.Arbitrum,
    name: 'Arbitrum',
    title: 'Arbitrum Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 42161,
    isTestnet: false,
    explorerUrl: 'https://arbiscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://arb1.arbitrum.io/rpc'
    ],
    eurcAddress: null,
    usdcAddress: '0xaf88d065e77c8cc2239327c5edb3a432268e5831',
    usdtAddress: null,
    cctp: {
        domain: 3,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x19330d10D9Cc8751218eaf51E8885D058642E08A',
                messageTransmitter: '0xC30362313FBBA5cf9163F0bb16a0e01f01A896ca',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 3,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Arbitrum Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the Arbitrum blockchain on Sepolia.
 */ const ArbitrumSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.Arbitrum_Sepolia,
    name: 'Arbitrum Sepolia',
    title: 'Arbitrum Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 421614,
    isTestnet: true,
    explorerUrl: 'https://sepolia.arbiscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://sepolia-rollup.arbitrum.io/rpc'
    ],
    eurcAddress: null,
    usdcAddress: '0x75faf114eafb1BDbe2F0316DF893fd58CE46AA4d',
    usdtAddress: null,
    cctp: {
        domain: 3,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9f3B8679c73C2Fef8b59B4f3444d4e156fb70AA5',
                messageTransmitter: '0xaCF1ceeF35caAc005e15888dDb8A3515C41B4872',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET,
        adapter: ADAPTER_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 3,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Avalanche Mainnet chain definition
 * @remarks
 * This represents the official production network for the Avalanche blockchain.
 */ const Avalanche = defineChain({
    type: 'evm',
    chain: Blockchain.Avalanche,
    name: 'Avalanche',
    title: 'Avalanche Mainnet',
    nativeCurrency: {
        name: 'Avalanche',
        symbol: 'AVAX',
        decimals: 18
    },
    chainId: 43114,
    isTestnet: false,
    explorerUrl: 'https://subnets.avax.network/c-chain/tx/{hash}',
    rpcEndpoints: [
        'https://api.avax.network/ext/bc/C/rpc'
    ],
    eurcAddress: '0xc891eb4cbdeff6e073e859e987815ed1505c2acd',
    usdcAddress: '0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E',
    usdtAddress: '0x9702230a8ea53601f5cd2dc00fdbc13d4df4a8c7',
    cctp: {
        domain: 1,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x6b25532e1060ce10cc3b0a99e5683b91bfde6982',
                messageTransmitter: '0x8186359af5f57fbb40c6b14a588d2a59c0c29880',
                confirmations: 1
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 1,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Avalanche Fuji Testnet chain definition
 * @remarks
 * This represents the official test network for the Avalanche blockchain.
 */ const AvalancheFuji = defineChain({
    type: 'evm',
    chain: Blockchain.Avalanche_Fuji,
    name: 'Avalanche Fuji',
    title: 'Avalanche Fuji Testnet',
    nativeCurrency: {
        name: 'Avalanche',
        symbol: 'AVAX',
        decimals: 18
    },
    chainId: 43113,
    isTestnet: true,
    explorerUrl: 'https://subnets-test.avax.network/c-chain/tx/{hash}',
    eurcAddress: '0x5e44db7996c682e92a960b65ac713a54ad815c6b',
    usdcAddress: '0x5425890298aed601595a70ab815c96711a31bc65',
    usdtAddress: null,
    cctp: {
        domain: 1,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0xeb08f243e5d3fcff26a9e38ae5520a669f4019d0',
                messageTransmitter: '0xa9fb1b3009dcb79e2fe346c16a604b8fa8ae0a79',
                confirmations: 1
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    rpcEndpoints: [
        'https://api.avax-test.network/ext/bc/C/rpc'
    ],
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 1,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Base chain definition
 * @remarks
 * This represents the official production network for the Base blockchain.
 */ const Base = defineChain({
    type: 'evm',
    chain: Blockchain.Base,
    name: 'Base',
    title: 'Base Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 8453,
    isTestnet: false,
    explorerUrl: 'https://basescan.org/tx/{hash}',
    rpcEndpoints: [
        'https://mainnet.base.org',
        'https://base.publicnode.com'
    ],
    eurcAddress: '0x60a3e35cc302bfa44cb288bc5a4f316fdb1adb42',
    usdcAddress: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
    usdtAddress: null,
    cctp: {
        domain: 6,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x1682Ae6375C4E4A97e4B583BC394c861A46D8962',
                messageTransmitter: '0xAD09780d193884d503182aD4588450C416D6F9D4',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 6,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Base Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the Base blockchain on Sepolia.
 */ const BaseSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.Base_Sepolia,
    name: 'Base Sepolia',
    title: 'Base Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 84532,
    isTestnet: true,
    explorerUrl: 'https://sepolia.basescan.org/tx/{hash}',
    rpcEndpoints: [
        'https://sepolia.base.org'
    ],
    eurcAddress: '0x808456652fdb597867f38412077A9182bf77359F',
    usdcAddress: '0x036CbD53842c5426634e7929541eC2318f3dCF7e',
    usdtAddress: null,
    cctp: {
        domain: 6,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9f3B8679c73C2Fef8b59B4f3444d4e156fb70AA5',
                messageTransmitter: '0x7865fAfC2db2093669d92c0F33AeEF291086BEFD',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET,
        adapter: ADAPTER_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 6,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Celo Mainnet chain definition
 * @remarks
 * This represents the official production network for the Celo blockchain.
 */ defineChain({
    type: 'evm',
    chain: Blockchain.Celo,
    name: 'Celo',
    title: 'Celo Mainnet',
    nativeCurrency: {
        name: 'Celo',
        symbol: 'CELO',
        decimals: 18
    },
    chainId: 42220,
    isTestnet: false,
    explorerUrl: 'https://celoscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://forno.celo.org'
    ],
    eurcAddress: null,
    usdcAddress: '0xcebA9300f2b948710d2653dD7B07f33A8B32118C',
    usdtAddress: '0x48065fbBE25f71C9282ddf5e1cD6D6A887483D5e',
    cctp: null
});

/**
 * Celo Alfajores Testnet chain definition
 * @remarks
 * This represents the official test network for the Celo blockchain.
 */ defineChain({
    type: 'evm',
    chain: Blockchain.Celo_Alfajores_Testnet,
    name: 'Celo Alfajores',
    title: 'Celo Alfajores Testnet',
    nativeCurrency: {
        name: 'Celo',
        symbol: 'CELO',
        decimals: 18
    },
    chainId: 44787,
    isTestnet: true,
    explorerUrl: 'https://alfajores.celoscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://alfajores-forno.celo-testnet.org'
    ],
    eurcAddress: null,
    usdcAddress: '0x2F25deB3848C207fc8E0c34035B3Ba7fC157602B',
    usdtAddress: null,
    cctp: null
});

/**
 * Codex Mainnet chain definition
 * @remarks
 * This represents the main network for the Codex blockchain.
 */ const Codex = defineChain({
    type: 'evm',
    chain: Blockchain.Codex,
    name: 'Codex Mainnet',
    title: 'Codex Mainnet',
    nativeCurrency: {
        name: 'ETH',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 81224,
    isTestnet: false,
    explorerUrl: 'https://explorer.codex.xyz/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.codex.xyz'
    ],
    eurcAddress: null,
    usdcAddress: '0xd996633a415985DBd7D6D12f4A4343E31f5037cf',
    usdtAddress: null,
    cctp: {
        domain: 12,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET
    }
});

/**
 * Codex Testnet chain definition
 * @remarks
 * This represents the test network for the Codex blockchain.
 */ const CodexTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Codex_Testnet,
    name: 'Codex Testnet',
    title: 'Codex Testnet',
    nativeCurrency: {
        name: 'ETH',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 812242,
    isTestnet: true,
    explorerUrl: 'https://explorer.codex-stg.xyz/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.codex-stg.xyz'
    ],
    eurcAddress: null,
    usdcAddress: '0x6d7f141b6819C2c9CC2f818e6ad549E7Ca090F8f',
    usdtAddress: null,
    cctp: {
        domain: 12,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Cronos Mainnet chain definition
 * @remarks
 * This represents the official production network for the Cronos blockchain.
 * Cronos is an EVM-compatible blockchain.
 */ const Cronos = defineChain({
    type: 'evm',
    chain: Blockchain.Cronos,
    name: 'Cronos',
    title: 'Cronos Mainnet',
    nativeCurrency: {
        name: 'Cronos',
        symbol: 'CRO',
        decimals: 18
    },
    chainId: 25,
    isTestnet: false,
    explorerUrl: 'https://cronoscan.com/tx/{hash}',
    rpcEndpoints: [
        'https://evm.cronos.org'
    ],
    eurcAddress: '0xA6dE01a2d62C6B5f3525d768f34d276652C554c8',
    usdcAddress: '0x3D7F2C478aAfdB65542BCB44bCeeC05849999d2D',
    usdtAddress: null,
    cctp: {
        domain: 32,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET
    }
});

/**
 * Cronos Testnet chain definition
 * @remarks
 * This represents the official test network for the Cronos blockchain.
 * Cronos is an EVM-compatible blockchain.
 */ const CronosTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Cronos_Testnet,
    name: 'Cronos Testnet',
    title: 'Cronos Testnet',
    nativeCurrency: {
        name: 'CRO',
        symbol: 'tCRO',
        decimals: 18
    },
    chainId: 338,
    isTestnet: true,
    explorerUrl: 'https://explorer.cronos.org/testnet/tx/{hash}',
    rpcEndpoints: [
        'https://evm-t3.cronos.org'
    ],
    eurcAddress: '0x31f7538adb53cF16350e6B0c89d03D91b7D12c46',
    usdcAddress: '0xEb33dc5fac03833e132593659e1dE7256aB59794',
    usdtAddress: null,
    cctp: {
        domain: 32,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Edge Mainnet chain definition
 * @remarks
 * This represents the official production network for the Edge blockchain.
 * Edge is an EVM-compatible blockchain.
 */ const Edge = defineChain({
    type: 'evm',
    chain: Blockchain.Edge,
    name: 'Edge',
    title: 'Edge Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 3343,
    isTestnet: false,
    explorerUrl: 'https://pro.edgex.exchange/en-US/explorer/tx/{hash}',
    rpcEndpoints: [
        'https://edge-mainnet.g.alchemy.com/public'
    ],
    eurcAddress: null,
    usdcAddress: '0x98d2919b9A214E6Fa5384AC81E6864bA686Ad74c',
    usdtAddress: null,
    cctp: {
        domain: 28,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x98706A006bc632Df31CAdFCBD43F38887ce2ca5c',
                messageTransmitter: '0x5b61381Fc9e58E70EfC13a4A97516997019198ee',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: '0x6D1AaE1c34Aeb582022916a67f2A655C6f4eDFF2'
    }
});

/**
 * Edge Testnet chain definition
 * @remarks
 * This represents the official test network for the Edge blockchain.
 * Edge is an EVM-compatible blockchain.
 */ const EdgeTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Edge_Testnet,
    name: 'Edge Testnet',
    title: 'Edge Testnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 33431,
    isTestnet: true,
    explorerUrl: 'https://edge-testnet.explorer.alchemy.com/tx/{hash}',
    rpcEndpoints: [
        'https://edge-testnet.g.alchemy.com/public'
    ],
    eurcAddress: null,
    usdcAddress: '0x2d9F7CAD728051AA35Ecdc472a14cf8cDF5CFD6B',
    usdtAddress: null,
    cctp: {
        domain: 28,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Ethereum Mainnet chain definition
 * @remarks
 * This represents the official production network for the Ethereum blockchain.
 */ const Ethereum = defineChain({
    type: 'evm',
    chain: Blockchain.Ethereum,
    name: 'Ethereum',
    title: 'Ethereum Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 1,
    isTestnet: false,
    explorerUrl: 'https://etherscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://ethereum-rpc.publicnode.com',
        'https://ethereum.publicnode.com'
    ],
    eurcAddress: '0x1aBaEA1f7C830bD89Acc67eC4af516284b1bC33c',
    usdcAddress: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
    usdtAddress: '0xdac17f958d2ee523a2206206994597c13d831ec7',
    cctp: {
        domain: 0,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0xbd3fa81b58ba92a82136038b25adec7066af3155',
                messageTransmitter: '0x0a992d191deec32afe36203ad87d7d289a738f81',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 2
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 0,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Ethereum Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the Ethereum blockchain on Sepolia.
 */ const EthereumSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.Ethereum_Sepolia,
    name: 'Ethereum Sepolia',
    title: 'Ethereum Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 11155111,
    isTestnet: true,
    explorerUrl: 'https://sepolia.etherscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://ethereum-sepolia-rpc.publicnode.com'
    ],
    eurcAddress: '0x08210F9170F89Ab7658F0B5E3fF39b0E03C594D4',
    usdcAddress: '0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238',
    usdtAddress: null,
    cctp: {
        domain: 0,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9f3B8679c73C2Fef8b59B4f3444d4e156fb70AA5',
                messageTransmitter: '0x7865fAfC2db2093669d92c0F33AeEF291086BEFD',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 65,
                fastConfirmations: 2
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET,
        adapter: ADAPTER_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 0,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Hedera Mainnet chain definition
 * @remarks
 * This represents the official production network for the Hedera blockchain.
 */ defineChain({
    type: 'hedera',
    chain: Blockchain.Hedera,
    name: 'Hedera',
    title: 'Hedera Mainnet',
    nativeCurrency: {
        name: 'HBAR',
        symbol: 'HBAR',
        decimals: 18
    },
    isTestnet: false,
    explorerUrl: 'https://hashscan.io/mainnet/transaction/{hash}',
    rpcEndpoints: [
        'https://mainnet.hashio.io/api'
    ],
    eurcAddress: null,
    usdcAddress: '0.0.456858',
    usdtAddress: null,
    cctp: null
});

/**
 * Hedera Testnet chain definition
 * @remarks
 * This represents the official test network for the Hedera blockchain.
 */ defineChain({
    type: 'hedera',
    chain: Blockchain.Hedera_Testnet,
    name: 'Hedera Testnet',
    title: 'Hedera Test Network',
    nativeCurrency: {
        name: 'HBAR',
        symbol: 'HBAR',
        decimals: 18
    },
    isTestnet: true,
    explorerUrl: 'https://hashscan.io/testnet/transaction/{hash}',
    rpcEndpoints: [
        'https://testnet.hashio.io/api'
    ],
    eurcAddress: null,
    usdcAddress: '0.0.429274',
    usdtAddress: null,
    cctp: null
});

/**
 * HyperEVM Mainnet chain definition
 * @remarks
 * This represents the official production network for the HyperEVM blockchain.
 * HyperEVM is a Layer 1 blockchain specialized for DeFi and trading applications
 * with native orderbook and matching engine.
 */ const HyperEVM = defineChain({
    type: 'evm',
    chain: Blockchain.HyperEVM,
    name: 'HyperEVM',
    title: 'HyperEVM Mainnet',
    nativeCurrency: {
        name: 'Hype',
        symbol: 'HYPE',
        decimals: 18
    },
    chainId: 999,
    isTestnet: false,
    explorerUrl: 'https://hyperevmscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.hyperliquid.xyz/evm'
    ],
    eurcAddress: null,
    usdcAddress: '0xb88339CB7199b77E23DB6E890353E22632Ba630f',
    usdtAddress: null,
    cctp: {
        domain: 19,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 19,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * HyperEVM Testnet chain definition
 * @remarks
 * This represents the official testnet for the HyperEVM blockchain.
 * Used for development and testing purposes before deploying to mainnet.
 */ const HyperEVMTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.HyperEVM_Testnet,
    name: 'HyperEVM Testnet',
    title: 'HyperEVM Test Network',
    nativeCurrency: {
        name: 'Hype',
        symbol: 'HYPE',
        decimals: 18
    },
    chainId: 998,
    isTestnet: true,
    explorerUrl: 'https://app.hyperliquid-testnet.xyz/explorer/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.hyperliquid-testnet.xyz/evm'
    ],
    eurcAddress: null,
    usdcAddress: '0x2B3370eE501B4a559b57D449569354196457D8Ab',
    usdtAddress: null,
    cctp: {
        domain: 19,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 19,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Injective Mainnet chain definition
 * @remarks
 * This represents the official production network for the Injective blockchain.
 * Injective is a high-performance, interoperable Layer-1 blockchain built for
 * finance, with an EVM execution layer on top of a Cosmos SDK base and
 * sub-second block finality.
 */ const Injective = defineChain({
    type: 'evm',
    chain: Blockchain.Injective,
    name: 'Injective',
    title: 'Injective Mainnet',
    nativeCurrency: {
        name: 'Injective',
        symbol: 'INJ',
        decimals: 18
    },
    chainId: 1776,
    isTestnet: false,
    explorerUrl: 'https://injscan.com/transaction/{hash}',
    rpcEndpoints: [
        'https://sentry.evm-rpc.injective.network'
    ],
    eurcAddress: null,
    usdcAddress: '0xa00C59fF5a080D2b954d0c75e46E22a0c371235a',
    usdtAddress: null,
    cctp: {
        domain: 29,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET
    }
});

/**
 * Injective Testnet chain definition
 * @remarks
 * This represents the official test network for the Injective blockchain.
 * Injective is a high-performance, interoperable Layer-1 blockchain built for
 * finance, with an EVM execution layer on top of a Cosmos SDK base and
 * sub-second block finality.
 */ const InjectiveTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Injective_Testnet,
    name: 'Injective Testnet',
    title: 'Injective Testnet',
    nativeCurrency: {
        name: 'Injective',
        symbol: 'INJ',
        decimals: 18
    },
    chainId: 1439,
    isTestnet: true,
    explorerUrl: 'https://testnet.explorer.injective.network/transaction/{hash}',
    rpcEndpoints: [
        'https://k8s.testnet.json-rpc.injective.network'
    ],
    eurcAddress: null,
    usdcAddress: '0x0C382e685bbeeFE5d3d9C29e29E341fEE8E84C5d',
    usdtAddress: null,
    cctp: {
        domain: 29,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Ink Mainnet chain definition
 * @remarks
 * This represents the official production network for the Ink blockchain.
 * Ink is a Layer 1 blockchain specialized for DeFi and trading applications
 * with native orderbook and matching engine.
 */ const Ink = defineChain({
    type: 'evm',
    chain: Blockchain.Ink,
    name: 'Ink',
    title: 'Ink Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 57073,
    isTestnet: false,
    explorerUrl: 'https://explorer.inkonchain.com/tx/{hash}',
    rpcEndpoints: [
        'https://rpc-gel.inkonchain.com',
        'https://rpc-qnd.inkonchain.com'
    ],
    eurcAddress: null,
    usdcAddress: '0x2D270e6886d130D724215A266106e6832161EAEd',
    usdtAddress: null,
    cctp: {
        domain: 21,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    }
});

/**
 * Ink Testnet chain definition
 * @remarks
 * This represents the official testnet for the Ink blockchain.
 * Used for development and testing purposes before deploying to mainnet.
 */ const InkTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Ink_Testnet,
    name: 'Ink Sepolia',
    title: 'Ink Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 763373,
    isTestnet: true,
    explorerUrl: 'https://explorer-sepolia.inkonchain.com/tx/{hash}',
    rpcEndpoints: [
        'https://rpc-gel-sepolia.inkonchain.com',
        'https://rpc-qnd-sepolia.inkonchain.com'
    ],
    eurcAddress: null,
    usdcAddress: '0xFabab97dCE620294D2B0b0e46C68964e326300Ac',
    usdtAddress: null,
    cctp: {
        domain: 21,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Linea Mainnet chain definition
 * @remarks
 * This represents the official production network for the Linea blockchain.
 */ const Linea = defineChain({
    type: 'evm',
    chain: Blockchain.Linea,
    name: 'Linea',
    title: 'Linea Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 59144,
    isTestnet: false,
    explorerUrl: 'https://lineascan.build/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.linea.build'
    ],
    eurcAddress: null,
    usdcAddress: '0x176211869ca2b568f2a7d4ee941e073a821ee1ff',
    usdtAddress: null,
    cctp: {
        domain: 11,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    }
});

/**
 * Linea Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the Linea blockchain on Sepolia.
 */ const LineaSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.Linea_Sepolia,
    name: 'Linea Sepolia',
    title: 'Linea Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 59141,
    isTestnet: true,
    explorerUrl: 'https://sepolia.lineascan.build/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.sepolia.linea.build'
    ],
    eurcAddress: null,
    usdcAddress: '0xfece4462d57bd51a6a552365a011b95f0e16d9b7',
    usdtAddress: null,
    cctp: {
        domain: 11,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Monad Mainnet chain definition
 * @remarks
 * This represents the official production network for the Monad blockchain.
 * Monad is a high-performance EVM-compatible Layer-1 blockchain featuring
 * over 10,000 TPS, sub-second finality, and near-zero gas fees.
 */ const Monad = defineChain({
    type: 'evm',
    chain: Blockchain.Monad,
    name: 'Monad',
    title: 'Monad Mainnet',
    nativeCurrency: {
        name: 'Monad',
        symbol: 'MON',
        decimals: 18
    },
    chainId: 143,
    isTestnet: false,
    explorerUrl: 'https://monadscan.com/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.monad.xyz'
    ],
    eurcAddress: null,
    usdcAddress: '0x754704Bc059F8C67012fEd69BC8A327a5aafb603',
    usdtAddress: null,
    cctp: {
        domain: 15,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    }
});

/**
 * Monad Testnet chain definition
 * @remarks
 * This represents the official test network for the Monad blockchain.
 * Monad is a high-performance EVM-compatible Layer-1 blockchain featuring
 * over 10,000 TPS, sub-second finality, and near-zero gas fees.
 */ const MonadTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Monad_Testnet,
    name: 'Monad Testnet',
    title: 'Monad Testnet',
    nativeCurrency: {
        name: 'Monad',
        symbol: 'MON',
        decimals: 18
    },
    chainId: 10143,
    isTestnet: true,
    explorerUrl: 'https://testnet.monadscan.com/tx/{hash}',
    rpcEndpoints: [
        'https://testnet-rpc.monad.xyz'
    ],
    eurcAddress: null,
    usdcAddress: '0x534b2f3A21130d7a60830c2Df862319e593943A3',
    usdtAddress: null,
    cctp: {
        domain: 15,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Morph Mainnet chain definition
 * @remarks
 * This represents the official production network for the Morph blockchain.
 * Morph is an EVM-compatible Layer-2 blockchain built on the OP Stack.
 */ const Morph = defineChain({
    type: 'evm',
    chain: Blockchain.Morph,
    name: 'Morph',
    title: 'Morph Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 2818,
    isTestnet: false,
    explorerUrl: 'https://explorer.morph.network/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.morphl2.io'
    ],
    eurcAddress: null,
    usdcAddress: '0xCfb1186F4e93D60E60a8bDd997427D1F33bc372B',
    usdtAddress: null,
    cctp: {
        domain: 30,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 64,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET
    }
});

/**
 * Morph Hoodi Testnet chain definition
 * @remarks
 * This represents the official test network for the Morph blockchain.
 * Morph is an EVM-compatible Layer-2 blockchain built on the OP Stack.
 */ const MorphTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Morph_Testnet,
    name: 'Morph Hoodi',
    title: 'Morph Hoodi Testnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 2910,
    isTestnet: true,
    explorerUrl: 'https://explorer-hoodi.morphl2.io/tx/{hash}',
    rpcEndpoints: [
        'https://rpc-hoodi.morphl2.io'
    ],
    eurcAddress: null,
    usdcAddress: '0x7433b41C6c5e1d58D4Da99483609520255ab661B',
    usdtAddress: null,
    cctp: {
        domain: 30,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 64,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * NEAR Protocol Mainnet chain definition
 * @remarks
 * This represents the official production network for the NEAR Protocol blockchain.
 */ defineChain({
    type: 'near',
    chain: Blockchain.NEAR,
    name: 'NEAR Protocol',
    title: 'NEAR Mainnet',
    nativeCurrency: {
        name: 'NEAR',
        symbol: 'NEAR',
        decimals: 24
    },
    isTestnet: false,
    explorerUrl: 'https://nearblocks.io/txns/{hash}',
    rpcEndpoints: [
        'https://eth-rpc.mainnet.near.org'
    ],
    eurcAddress: null,
    usdcAddress: '17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1',
    usdtAddress: 'usdt.tether-token.near',
    cctp: null
});

/**
 * NEAR Testnet chain definition
 * @remarks
 * This represents the official test network for the NEAR Protocol blockchain.
 */ defineChain({
    type: 'near',
    chain: Blockchain.NEAR_Testnet,
    name: 'NEAR Protocol Testnet',
    title: 'NEAR Test Network',
    nativeCurrency: {
        name: 'NEAR',
        symbol: 'NEAR',
        decimals: 24
    },
    isTestnet: true,
    explorerUrl: 'https://testnet.nearblocks.io/txns/{hash}',
    rpcEndpoints: [
        'https://eth-rpc.testnet.near.org'
    ],
    eurcAddress: null,
    usdcAddress: '3e2210e1184b45b64c8a434c0a7e7b23cc04ea7eb7a6c3c32520d03d4afcb8af',
    usdtAddress: null,
    cctp: null
});

/**
 * Noble Mainnet chain definition
 * @remarks
 * This represents the official production network for the Noble blockchain.
 */ defineChain({
    type: 'noble',
    chain: Blockchain.Noble,
    name: 'Noble',
    title: 'Noble Mainnet',
    nativeCurrency: {
        name: 'Noble USDC',
        symbol: 'USDC',
        decimals: 6
    },
    isTestnet: false,
    explorerUrl: 'https://www.mintscan.io/noble/tx/{hash}',
    rpcEndpoints: [
        'https://noble-rpc.polkachu.com'
    ],
    eurcAddress: null,
    usdcAddress: 'uusdc',
    usdtAddress: null,
    cctp: {
        domain: 4,
        contracts: {
            v1: {
                type: 'merged',
                contract: 'noble12l2w4ugfz4m6dd73yysz477jszqnfughxvkss5',
                confirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    }
});

/**
 * Noble Testnet chain definition
 * @remarks
 * This represents the official test network for the Noble blockchain.
 */ defineChain({
    type: 'noble',
    chain: Blockchain.Noble_Testnet,
    name: 'Noble Testnet',
    title: 'Noble Test Network',
    nativeCurrency: {
        name: 'Noble USDC',
        symbol: 'USDC',
        decimals: 6
    },
    isTestnet: true,
    explorerUrl: 'https://www.mintscan.io/noble-testnet/tx/{hash}',
    rpcEndpoints: [
        'https://noble-testnet-rpc.polkachu.com'
    ],
    eurcAddress: null,
    usdcAddress: 'uusdc',
    usdtAddress: null,
    cctp: {
        domain: 4,
        contracts: {
            v1: {
                type: 'merged',
                contract: 'noble12l2w4ugfz4m6dd73yysz477jszqnfughxvkss5',
                confirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    }
});

/**
 * Optimism Mainnet chain definition
 * @remarks
 * This represents the official production network for the Optimism blockchain.
 */ const Optimism = defineChain({
    type: 'evm',
    chain: Blockchain.Optimism,
    name: 'Optimism',
    title: 'Optimism Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 10,
    isTestnet: false,
    explorerUrl: 'https://optimistic.etherscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://mainnet.optimism.io'
    ],
    eurcAddress: null,
    usdcAddress: '0x0b2c639c533813f4aa9d7837caf62653d097ff85',
    usdtAddress: null,
    cctp: {
        domain: 2,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x2B4069517957735bE00ceE0fadAE88a26365528f',
                messageTransmitter: '0x0a992d191deec32afe36203ad87d7d289a738f81',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 2,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Optimism Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the Optimism blockchain on Sepolia.
 */ const OptimismSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.Optimism_Sepolia,
    name: 'Optimism Sepolia',
    title: 'Optimism Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 11155420,
    isTestnet: true,
    explorerUrl: 'https://sepolia-optimistic.etherscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://sepolia.optimism.io'
    ],
    eurcAddress: null,
    usdcAddress: '0x5fd84259d66Cd46123540766Be93DFE6D43130D7',
    usdtAddress: null,
    cctp: {
        domain: 2,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9f3B8679c73C2Fef8b59B4f3444d4e156fb70AA5',
                messageTransmitter: '0x7865fAfC2db2093669d92c0F33AeEF291086BEFD',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 2,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Pharos Mainnet chain definition
 * @remarks
 * This represents the official production network for the Pharos blockchain.
 * Pharos is a modular, full-stack parallel Layer 1 blockchain with
 * sub-second finality and EVM compatibility.
 */ const Pharos = defineChain({
    type: 'evm',
    chain: Blockchain.Pharos,
    name: 'Pharos',
    title: 'Pharos Mainnet',
    nativeCurrency: {
        name: 'Pharos',
        symbol: 'PHAROS',
        decimals: 18
    },
    chainId: 1672,
    isTestnet: false,
    explorerUrl: 'https://pharos.socialscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.pharos.xyz'
    ],
    eurcAddress: null,
    usdcAddress: '0xC879C018dB60520F4355C26eD1a6D572cdAC1815',
    usdtAddress: null,
    cctp: {
        domain: 31,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET
    }
});

/**
 * Pharos Atlantic Testnet chain definition
 * @remarks
 * This represents the official test network for the Pharos blockchain.
 * Pharos is a modular, full-stack parallel Layer 1 blockchain with
 * sub-second finality and EVM compatibility.
 */ const PharosTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Pharos_Testnet,
    name: 'Pharos Atlantic',
    title: 'Pharos Atlantic Testnet',
    nativeCurrency: {
        name: 'Pharos',
        symbol: 'PHAROS',
        decimals: 18
    },
    chainId: 688689,
    isTestnet: true,
    explorerUrl: 'https://atlantic.pharosscan.xyz/tx/{hash}',
    rpcEndpoints: [
        'https://atlantic.dplabs-internal.com'
    ],
    eurcAddress: null,
    usdcAddress: '0xcfC8330f4BCAB529c625D12781b1C19466A9Fc8B',
    usdtAddress: null,
    cctp: {
        domain: 31,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Plume Mainnet chain definition
 * @remarks
 * This represents the official production network for the Plume blockchain.
 * Plume is a Layer 1 blockchain specialized for DeFi and trading applications
 * with native orderbook and matching engine.
 */ const Plume = defineChain({
    type: 'evm',
    chain: Blockchain.Plume,
    name: 'Plume',
    title: 'Plume Mainnet',
    nativeCurrency: {
        name: 'Plume',
        symbol: 'PLUME',
        decimals: 18
    },
    chainId: 98866,
    isTestnet: false,
    explorerUrl: 'https://explorer.plume.org/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.plume.org'
    ],
    eurcAddress: null,
    usdcAddress: '0x222365EF19F7947e5484218551B56bb3965Aa7aF',
    usdtAddress: null,
    cctp: {
        domain: 22,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    }
});

/**
 * Plume Testnet chain definition
 * @remarks
 * This represents the official testnet for the Plume blockchain.
 * Used for development and testing purposes before deploying to mainnet.
 */ const PlumeTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Plume_Testnet,
    name: 'Plume Testnet',
    title: 'Plume Test Network',
    nativeCurrency: {
        name: 'Plume',
        symbol: 'PLUME',
        decimals: 18
    },
    chainId: 98867,
    isTestnet: true,
    explorerUrl: 'https://testnet-explorer.plume.org/tx/{hash}',
    rpcEndpoints: [
        'https://testnet-rpc.plume.org'
    ],
    eurcAddress: null,
    usdcAddress: '0xcB5f30e335672893c7eb944B374c196392C19D18',
    usdtAddress: null,
    cctp: {
        domain: 22,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * Polkadot Asset Hub chain definition
 * @remarks
 * This represents the official asset management parachain for the Polkadot blockchain.
 */ defineChain({
    type: 'polkadot',
    chain: Blockchain.Polkadot_Asset_Hub,
    name: 'Polkadot Asset Hub',
    title: 'Polkadot Asset Hub',
    nativeCurrency: {
        name: 'Polkadot',
        symbol: 'DOT',
        decimals: 10
    },
    isTestnet: false,
    explorerUrl: 'https://polkadot.subscan.io/extrinsic/{hash}',
    rpcEndpoints: [
        'https://asset-hub-polkadot-rpc.n.dwellir.com'
    ],
    eurcAddress: null,
    usdcAddress: '1337',
    usdtAddress: '1984',
    cctp: null
});

/**
 * Polkadot Westmint chain definition
 * @remarks
 * This represents an asset management parachain in the Polkadot ecosystem.
 */ defineChain({
    type: 'polkadot',
    chain: Blockchain.Polkadot_Westmint,
    name: 'Polkadot Westmint',
    title: 'Polkadot Westmint',
    nativeCurrency: {
        name: 'Polkadot',
        symbol: 'DOT',
        decimals: 10
    },
    isTestnet: false,
    explorerUrl: 'https://assethub-polkadot.subscan.io/extrinsic/{hash}',
    rpcEndpoints: [
        'https://westmint-rpc.polkadot.io'
    ],
    eurcAddress: null,
    usdcAddress: 'Asset ID 31337',
    usdtAddress: null,
    cctp: null
});

/**
 * Polygon Mainnet chain definition
 * @remarks
 * This represents the official production network for the Polygon blockchain.
 */ const Polygon = defineChain({
    type: 'evm',
    chain: Blockchain.Polygon,
    name: 'Polygon',
    title: 'Polygon Mainnet',
    nativeCurrency: {
        name: 'POL',
        symbol: 'POL',
        decimals: 18
    },
    chainId: 137,
    isTestnet: false,
    explorerUrl: 'https://polygonscan.com/tx/{hash}',
    rpcEndpoints: [
        'https://polygon.publicnode.com',
        'https://polygon.drpc.org'
    ],
    eurcAddress: null,
    usdcAddress: '0x3c499c542cef5e3811e1192ce70d8cc03d5c3359',
    usdtAddress: null,
    cctp: {
        domain: 7,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9daF8c91AEFAE50b9c0E69629D3F6Ca40cA3B3FE',
                messageTransmitter: '0xF3be9355363857F3e001be68856A2f96b4C39Ba9',
                confirmations: 200
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 33,
                fastConfirmations: 13
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 7,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Polygon Amoy Testnet chain definition
 * @remarks
 * This represents the official test network for the Polygon blockchain.
 */ const PolygonAmoy = defineChain({
    type: 'evm',
    chain: Blockchain.Polygon_Amoy_Testnet,
    name: 'Polygon Amoy',
    title: 'Polygon Amoy Testnet',
    nativeCurrency: {
        name: 'POL',
        symbol: 'POL',
        decimals: 18
    },
    chainId: 80002,
    isTestnet: true,
    explorerUrl: 'https://amoy.polygonscan.com/tx/{hash}',
    rpcEndpoints: [
        'https://rpc-amoy.polygon.technology'
    ],
    eurcAddress: null,
    usdcAddress: '0x41e94eb019c0762f9bfcf9fb1e58725bfb0e7582',
    usdtAddress: null,
    cctp: {
        domain: 7,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x9f3B8679c73C2Fef8b59B4f3444d4e156fb70AA5',
                messageTransmitter: '0x7865fAfC2db2093669d92c0F33AeEF291086BEFD',
                confirmations: 200
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 33,
                fastConfirmations: 13
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 7,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Sei Mainnet chain definition
 * @remarks
 * This represents the official production network for the Sei blockchain.
 * Sei is a Layer 1 blockchain specialized for DeFi and trading applications
 * with native orderbook and matching engine.
 */ const Sei = defineChain({
    type: 'evm',
    chain: Blockchain.Sei,
    name: 'Sei',
    title: 'Sei Mainnet',
    nativeCurrency: {
        name: 'Sei',
        symbol: 'SEI',
        decimals: 18
    },
    chainId: 1329,
    isTestnet: false,
    explorerUrl: 'https://seiscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://evm-rpc.sei-apis.com'
    ],
    eurcAddress: null,
    usdcAddress: '0xe15fC38F6D8c56aF07bbCBe3BAf5708A2Bf42392',
    usdtAddress: null,
    cctp: {
        domain: 16,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 16,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Sei Testnet chain definition
 * @remarks
 * This represents the official testnet for the Sei blockchain.
 * Used for development and testing purposes before deploying to mainnet.
 */ const SeiTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Sei_Testnet,
    name: 'Sei Testnet',
    title: 'Sei Test Network',
    nativeCurrency: {
        name: 'Sei',
        symbol: 'SEI',
        decimals: 18
    },
    chainId: 1328,
    isTestnet: true,
    explorerUrl: 'https://testnet.seiscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://evm-rpc-testnet.sei-apis.com'
    ],
    eurcAddress: null,
    usdcAddress: '0x4fCF1784B31630811181f670Aea7A7bEF803eaED',
    usdtAddress: null,
    cctp: {
        domain: 16,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 16,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Sonic Mainnet chain definition
 * @remarks
 * This represents the official production network for the Sonic blockchain.
 */ const Sonic = defineChain({
    type: 'evm',
    chain: Blockchain.Sonic,
    name: 'Sonic',
    title: 'Sonic Mainnet',
    nativeCurrency: {
        name: 'Sonic',
        symbol: 'S',
        decimals: 18
    },
    chainId: 146,
    isTestnet: false,
    explorerUrl: 'https://sonicscan.org/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.soniclabs.com'
    ],
    eurcAddress: null,
    usdcAddress: '0x29219dd400f2Bf60E5a23d13Be72B486D4038894',
    usdtAddress: null,
    cctp: {
        domain: 13,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 13,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Sonic Testnet chain definition
 * @remarks
 * This represents the official test network for the Sonic blockchain.
 */ const SonicTestnet = defineChain({
    type: 'evm',
    chain: Blockchain.Sonic_Testnet,
    name: 'Sonic Testnet',
    title: 'Sonic Testnet',
    nativeCurrency: {
        name: 'Sonic',
        symbol: 'S',
        decimals: 18
    },
    chainId: 14601,
    isTestnet: true,
    explorerUrl: 'https://testnet.sonicscan.org/tx/{hash}',
    rpcEndpoints: [
        'https://rpc.testnet.soniclabs.com'
    ],
    eurcAddress: null,
    usdcAddress: '0x0BA304580ee7c9a980CF72e55f5Ed2E9fd30Bc51',
    usdtAddress: null,
    cctp: {
        domain: 13,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 1,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 13,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Solana Mainnet chain definition
 * @remarks
 * This represents the official production network for the Solana blockchain.
 */ const Solana = defineChain({
    type: 'solana',
    chain: Blockchain.Solana,
    name: 'Solana',
    title: 'Solana Mainnet',
    nativeCurrency: {
        name: 'Solana',
        symbol: 'SOL',
        decimals: 9
    },
    isTestnet: false,
    explorerUrl: 'https://solscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://api.mainnet-beta.solana.com'
    ],
    eurcAddress: 'HzwqbKZw8HxMN6bF2yFZNrht3c2iXXzpKcFu7uBEDKtr',
    usdcAddress: 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v',
    usdtAddress: 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB',
    cctp: {
        domain: 5,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: 'CCTPiPYPc6AsJuwueEnWgSgucamXDZwBd53dQ11YiKX3',
                messageTransmitter: 'CCTPmbSD7gX1bxKPAmg77w8oFzNFpaQiQUWD43TKaecd',
                confirmations: 32
            },
            v2: {
                type: 'split',
                tokenMessenger: 'CCTPV2vPZJS2u2BBsUoscuikbYjnpFmbFsvVuJdgUMQe',
                messageTransmitter: 'CCTPV2Sm4AdWt5296sk4P66VBZ7bEhcARwFaaS9YPbeC',
                confirmations: 32,
                fastConfirmations: 3
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: 'DFaauJEjmiHkPs1JG89A4p95hDWi9m9SAEERY1LQJiC3'
    },
    gateway: {
        domain: 5,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_SOLANA_MAINNET,
                minter: GATEWAY_MINTER_SOLANA_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Solana Devnet chain definition
 * @remarks
 * This represents the development test network for the Solana blockchain.
 */ const SolanaDevnet = defineChain({
    type: 'solana',
    chain: Blockchain.Solana_Devnet,
    name: 'Solana Devnet',
    title: 'Solana Development Network',
    nativeCurrency: {
        name: 'Solana',
        symbol: 'SOL',
        decimals: 9
    },
    isTestnet: true,
    explorerUrl: 'https://solscan.io/tx/{hash}?cluster=devnet',
    eurcAddress: 'HzwqbKZw8HxMN6bF2yFZNrht3c2iXXzpKcFu7uBEDKtr',
    usdcAddress: '4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU',
    usdtAddress: null,
    cctp: {
        domain: 5,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: 'CCTPiPYPc6AsJuwueEnWgSgucamXDZwBd53dQ11YiKX3',
                messageTransmitter: 'CCTPmbSD7gX1bxKPAmg77w8oFzNFpaQiQUWD43TKaecd',
                confirmations: 32
            },
            v2: {
                type: 'split',
                tokenMessenger: 'CCTPV2vPZJS2u2BBsUoscuikbYjnpFmbFsvVuJdgUMQe',
                messageTransmitter: 'CCTPV2Sm4AdWt5296sk4P66VBZ7bEhcARwFaaS9YPbeC',
                confirmations: 32,
                fastConfirmations: 3
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: 'DFaauJEjmiHkPs1JG89A4p95hDWi9m9SAEERY1LQJiC3'
    },
    rpcEndpoints: [
        'https://api.devnet.solana.com'
    ],
    gateway: {
        domain: 5,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_SOLANA_DEVNET,
                minter: GATEWAY_MINTER_SOLANA_DEVNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Stellar Mainnet chain definition
 * @remarks
 * This represents the official production network for the Stellar blockchain.
 */ defineChain({
    type: 'stellar',
    chain: Blockchain.Stellar,
    name: 'Stellar',
    title: 'Stellar Mainnet',
    nativeCurrency: {
        name: 'Stellar Lumens',
        symbol: 'XLM',
        decimals: 7
    },
    isTestnet: false,
    explorerUrl: 'https://stellar.expert/explorer/public/tx/{hash}',
    rpcEndpoints: [
        'https://horizon.stellar.org'
    ],
    eurcAddress: 'EURC-GDHU6WRG4IEQXM5NZ4BMPKOXHW76MZM4Y2IEMFDVXBSDP6SJY4ITNPP2',
    usdcAddress: 'USDC-GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN',
    usdtAddress: null,
    cctp: null
});

/**
 * Stellar Testnet chain definition
 * @remarks
 * This represents the official test network for the Stellar blockchain.
 */ defineChain({
    type: 'stellar',
    chain: Blockchain.Stellar_Testnet,
    name: 'Stellar Testnet',
    title: 'Stellar Test Network',
    nativeCurrency: {
        name: 'Stellar Lumens',
        symbol: 'XLM',
        decimals: 7
    },
    isTestnet: true,
    explorerUrl: 'https://stellar.expert/explorer/testnet/tx/{hash}',
    rpcEndpoints: [
        'https://horizon-testnet.stellar.org'
    ],
    eurcAddress: 'EURC-GB3Q6QDZYTHWT7E5PVS3W7FUT5GVAFC5KSZFFLPU25GO7VTC3NM2ZTVO',
    usdcAddress: 'USDC-GBBD47IF6LWK7P7MDEVSCWR7DPUWV3NY3DTQEVFL4NAT4AQH3ZLLFLA5',
    usdtAddress: null,
    cctp: null
});

/**
 * Sui Mainnet chain definition
 * @remarks
 * This represents the official production network for the Sui blockchain.
 */ defineChain({
    type: 'sui',
    chain: Blockchain.Sui,
    name: 'Sui',
    title: 'Sui Mainnet',
    nativeCurrency: {
        name: 'Sui',
        symbol: 'SUI',
        decimals: 9
    },
    isTestnet: false,
    explorerUrl: 'https://suiscan.xyz/mainnet/tx/{hash}',
    rpcEndpoints: [
        'https://fullnode.mainnet.sui.io'
    ],
    eurcAddress: null,
    usdcAddress: '0xdba34672e30cb065b1f93e3ab55318768fd6fef66c15942c9f7cb846e2f900e7::usdc::USDC',
    usdtAddress: null,
    cctp: {
        domain: 8,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x2aa6c5d56376c371f88a6cc42e852824994993cb9bab8d3e6450cbe3cb32b94e',
                messageTransmitter: '0x08d87d37ba49e785dde270a83f8e979605b03dc552b5548f26fdf2f49bf7ed1b',
                confirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    }
});

/**
 * Sui Testnet chain definition
 * @remarks
 * This represents the official test network for the Sui blockchain.
 */ defineChain({
    type: 'sui',
    chain: Blockchain.Sui_Testnet,
    name: 'Sui Testnet',
    title: 'Sui Test Network',
    nativeCurrency: {
        name: 'Sui',
        symbol: 'SUI',
        decimals: 9
    },
    isTestnet: true,
    explorerUrl: 'https://suiscan.xyz/testnet/tx/{hash}',
    rpcEndpoints: [
        'https://fullnode.testnet.sui.io'
    ],
    eurcAddress: null,
    usdcAddress: '0xa1ec7fc00a6f40db9693ad1415d0c193ad3906494428cf252621037bd7117e29::usdc::USDC',
    usdtAddress: null,
    cctp: {
        domain: 8,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x31cc14d80c175ae39777c0238f20594c6d4869cfab199f40b69f3319956b8beb',
                messageTransmitter: '0x4931e06dce648b3931f890035bd196920770e913e43e45990b383f6486fdd0a5',
                confirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: false
        }
    }
});

/**
 * Unichain Mainnet chain definition
 * @remarks
 * This represents the official production network for the Unichain blockchain.
 */ const Unichain = defineChain({
    type: 'evm',
    chain: Blockchain.Unichain,
    name: 'Unichain',
    title: 'Unichain Mainnet',
    nativeCurrency: {
        name: 'Uni',
        symbol: 'UNI',
        decimals: 18
    },
    chainId: 130,
    isTestnet: false,
    explorerUrl: 'https://unichain.blockscout.com/tx/{hash}',
    rpcEndpoints: [
        'https://mainnet.unichain.org'
    ],
    eurcAddress: null,
    usdcAddress: '0x078D782b760474a361dDA0AF3839290b0EF57AD6',
    usdtAddress: null,
    cctp: {
        domain: 10,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x4e744b28E787c3aD0e810eD65A24461D4ac5a762',
                messageTransmitter: '0x353bE9E2E38AB1D19104534e4edC21c643Df86f4',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 10,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * Unichain Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the Unichain blockchain.
 */ const UnichainSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.Unichain_Sepolia,
    name: 'Unichain Sepolia',
    title: 'Unichain Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Uni',
        symbol: 'UNI',
        decimals: 18
    },
    chainId: 1301,
    isTestnet: true,
    explorerUrl: 'https://unichain-sepolia.blockscout.com/tx/{hash}',
    rpcEndpoints: [
        'https://sepolia.unichain.org'
    ],
    eurcAddress: null,
    usdcAddress: '0x31d0220469e10c4E71834a79b1f276d740d3768F',
    usdtAddress: null,
    cctp: {
        domain: 10,
        contracts: {
            v1: {
                type: 'split',
                tokenMessenger: '0x8ed94B8dAd2Dc5453862ea5e316A8e71AAed9782',
                messageTransmitter: '0xbc498c326533d675cf571B90A2Ced265ACb7d086',
                confirmations: 65
            },
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 10,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * World Chain chain definition
 * @remarks
 * This represents the main network for the World Chain blockchain.
 */ const WorldChain = defineChain({
    type: 'evm',
    chain: Blockchain.World_Chain,
    name: 'World Chain',
    title: 'World Chain',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 480,
    isTestnet: false,
    explorerUrl: 'https://worldscan.org/tx/{hash}',
    rpcEndpoints: [
        'https://worldchain-mainnet.g.alchemy.com/public'
    ],
    eurcAddress: null,
    usdcAddress: '0x79A02482A880bCE3F13e09Da970dC34db4CD24d1',
    usdtAddress: null,
    cctp: {
        domain: 14,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cF5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    },
    gateway: {
        domain: 14,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_MAINNET,
                minter: GATEWAY_MINTER_EVM_MAINNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * World Chain Sepolia chain definition
 * @remarks
 * This represents the test network for the World Chain blockchain.
 */ const WorldChainSepolia = defineChain({
    type: 'evm',
    chain: Blockchain.World_Chain_Sepolia,
    name: 'World Chain Sepolia',
    title: 'World Chain Sepolia',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 4801,
    isTestnet: true,
    explorerUrl: 'https://sepolia.worldscan.org/tx/{hash}',
    rpcEndpoints: [
        'https://worldchain-sepolia.drpc.org',
        'https://worldchain-sepolia.g.alchemy.com/public'
    ],
    eurcAddress: null,
    usdcAddress: '0x66145f38cBAC35Ca6F1Dfb4914dF98F1614aeA88',
    usdtAddress: null,
    cctp: {
        domain: 14,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8fe6b999dc680ccfdd5bf7eb0974218be2542daa',
                messageTransmitter: '0xe737e5cebeeba77efe34d4aa090756590b1ce275',
                confirmations: 65,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    },
    gateway: {
        domain: 14,
        contracts: {
            v1: {
                wallet: GATEWAY_WALLET_EVM_TESTNET,
                minter: GATEWAY_MINTER_EVM_TESTNET
            }
        },
        forwarderSupported: {
            source: true,
            destination: true
        }
    }
});

/**
 * XDC Mainnet chain definition
 * @remarks
 * This represents the official production network for the XDC blockchain.
 * XDC is a Layer 1 blockchain specialized for DeFi and trading applications
 * with native orderbook and matching engine.
 */ const XDC = defineChain({
    type: 'evm',
    chain: Blockchain.XDC,
    name: 'XDC',
    title: 'XDC Mainnet',
    nativeCurrency: {
        name: 'XDC',
        symbol: 'XDC',
        decimals: 18
    },
    chainId: 50,
    isTestnet: false,
    explorerUrl: 'https://xdcscan.io/tx/{hash}',
    rpcEndpoints: [
        'https://erpc.xdcrpc.com',
        'https://erpc.xinfin.network'
    ],
    eurcAddress: null,
    usdcAddress: '0xfA2958CB79b0491CC627c1557F441eF849Ca8eb1',
    usdtAddress: null,
    cctp: {
        domain: 18,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x28b5a0e9C621a5BadaA536219b3a228C8168cf5d',
                messageTransmitter: '0x81D40F21F12A8F0E3252Bccb954D722d4c464B64',
                confirmations: 3,
                fastConfirmations: 3
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_MAINNET,
        adapter: ADAPTER_CONTRACT_EVM_MAINNET
    }
});

/**
 * XDC Apothem Testnet chain definition
 * @remarks
 * This represents the official test network for the XDC Network, known as Apothem.
 */ const XDCApothem = defineChain({
    type: 'evm',
    chain: Blockchain.XDC_Apothem,
    name: 'Apothem Network',
    title: 'Apothem Network',
    nativeCurrency: {
        name: 'TXDC',
        symbol: 'TXDC',
        decimals: 18
    },
    chainId: 51,
    isTestnet: true,
    explorerUrl: 'https://testnet.xdcscan.com/tx/{hash}',
    rpcEndpoints: [
        'https://erpc.apothem.network'
    ],
    eurcAddress: null,
    usdcAddress: '0xb5AB69F7bBada22B28e79C8FFAECe55eF1c771D4',
    usdtAddress: null,
    cctp: {
        domain: 18,
        contracts: {
            v2: {
                type: 'split',
                tokenMessenger: '0x8FE6B999Dc680CcFDD5Bf7EB0974218be2542DAA',
                messageTransmitter: '0xE737e5cEBEEBa77EFE34D4aa090756590b1CE275',
                confirmations: 3,
                fastConfirmations: 1
            }
        },
        forwarderSupported: {
            source: false,
            destination: true
        }
    },
    kitContracts: {
        bridge: BRIDGE_CONTRACT_EVM_TESTNET
    }
});

/**
 * ZKSync Era Mainnet chain definition
 * @remarks
 * This represents the official production network for the ZKSync Era blockchain.
 */ defineChain({
    type: 'evm',
    chain: Blockchain.ZKSync_Era,
    name: 'ZKSync Era',
    title: 'ZKSync Era Mainnet',
    nativeCurrency: {
        name: 'Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 324,
    isTestnet: false,
    explorerUrl: 'https://explorer.zksync.io/tx/{hash}',
    rpcEndpoints: [
        'https://mainnet.era.zksync.io'
    ],
    eurcAddress: null,
    usdcAddress: '0x1d17CBcF0D6D143135aE902365D2E5e2A16538D4',
    usdtAddress: null,
    cctp: null
});

/**
 * ZKSync Era Sepolia Testnet chain definition
 * @remarks
 * This represents the official test network for the ZKSync Era blockchain on Sepolia.
 */ defineChain({
    type: 'evm',
    chain: Blockchain.ZKSync_Sepolia,
    name: 'ZKSync Era Sepolia',
    title: 'ZKSync Era Sepolia Testnet',
    nativeCurrency: {
        name: 'Sepolia Ether',
        symbol: 'ETH',
        decimals: 18
    },
    chainId: 300,
    isTestnet: true,
    explorerUrl: 'https://sepolia.explorer.zksync.io/tx/{hash}',
    rpcEndpoints: [
        'https://sepolia.era.zksync.dev'
    ],
    eurcAddress: null,
    usdcAddress: '0xAe045DE5638162fa134807Cb558E15A3F5A7F853',
    usdtAddress: null,
    cctp: null
});

/**
 * Zod schema for validating Gateway v1 contract addresses.
 *
 * @example
 * ```typescript
 * gatewayV1ContractsSchema.parse({
 *   wallet: '0x1234567890abcdef1234567890abcdef12345678',
 *   minter: '0xabcdef1234567890abcdef1234567890abcdef12'
 * })
 * ```
 */ const gatewayV1ContractsSchema = zod.z.object({
    wallet: zod.z.string({
        required_error: 'Gateway wallet address is required. Please provide a valid contract address.',
        invalid_type_error: 'Gateway wallet address must be a string.'
    }).min(1, 'Gateway wallet address cannot be empty.'),
    minter: zod.z.string({
        required_error: 'Gateway minter address is required. Please provide a valid contract address.',
        invalid_type_error: 'Gateway minter address must be a string.'
    }).min(1, 'Gateway minter address cannot be empty.'),
    depositForHandler: zod.z.string({
        invalid_type_error: 'Gateway depositForHandler address must be a string.'
    }).min(1, 'Gateway depositForHandler address cannot be empty.').optional()
}).strict() // Reject any additional properties not defined in the schema
;
/**
 * Zod schema for validating the versioned Gateway contracts map.
 *
 * @description Mirrors the {@link GatewayContracts} type: a partial map of
 * protocol versions to their contract addresses, following the same pattern
 * as {@link CCTPContracts}.
 *
 * @example
 * ```typescript
 * gatewayContractsSchema.parse({
 *   v1: {
 *     wallet: '0x1234567890abcdef1234567890abcdef12345678',
 *     minter: '0xabcdef1234567890abcdef1234567890abcdef12'
 *   }
 * })
 * ```
 */ const gatewayContractsSchema = zod.z.object({
    v1: gatewayV1ContractsSchema.optional()
}).strict() // Reject any additional properties not defined in the schema
;
/**
 * Zod schema for validating the full Gateway configuration.
 *
 * @description Mirrors the {@link GatewayConfig} type: a domain number plus
 * a versioned contracts map, following the same pattern as {@link CCTPConfig}.
 *
 * @example
 * ```typescript
 * gatewayConfigSchema.parse({
 *   domain: 6,
 *   contracts: {
 *     v1: {
 *       wallet: '0x1234567890abcdef1234567890abcdef12345678',
 *       minter: '0xabcdef1234567890abcdef1234567890abcdef12'
 *     }
 *   }
 * })
 * ```
 */ const gatewayConfigSchema = zod.z.object({
    domain: zod.z.number({
        required_error: 'Gateway domain is required. Please provide a valid domain number.',
        invalid_type_error: 'Gateway domain must be a number.'
    }),
    contracts: gatewayContractsSchema,
    forwarderSupported: zod.z.object({
        source: zod.z.boolean(),
        destination: zod.z.boolean()
    })
}).strict() // Reject any additional properties not defined in the schema
;
/**
 * Base schema for common chain definition properties.
 * This contains all properties shared between EVM and non-EVM chains.
 */ const baseChainDefinitionSchema = zod.z.object({
    chain: zod.z.nativeEnum(Blockchain, {
        required_error: 'Chain enum is required. Please provide a valid Blockchain enum value.',
        invalid_type_error: 'Chain must be a valid Blockchain enum value.'
    }),
    name: zod.z.string({
        required_error: 'Chain name is required. Please provide a valid chain name.',
        invalid_type_error: 'Chain name must be a string.'
    }),
    title: zod.z.string().optional(),
    nativeCurrency: zod.z.object({
        name: zod.z.string(),
        symbol: zod.z.string(),
        decimals: zod.z.number()
    }),
    isTestnet: zod.z.boolean({
        required_error: 'isTestnet is required. Please specify whether this is a testnet.',
        invalid_type_error: 'isTestnet must be a boolean.'
    }),
    explorerUrl: zod.z.string({
        required_error: 'Explorer URL is required. Please provide a valid explorer URL.',
        invalid_type_error: 'Explorer URL must be a string.'
    }),
    rpcEndpoints: zod.z.array(zod.z.string()),
    eurcAddress: zod.z.string().nullable(),
    usdcAddress: zod.z.string().nullable(),
    usdtAddress: zod.z.string().nullable(),
    cctp: zod.z.any().nullable(),
    kitContracts: zod.z.object({
        bridge: zod.z.string().optional(),
        adapter: zod.z.string().optional()
    }).optional(),
    gateway: gatewayConfigSchema.optional()
});
/**
 * Zod schema for validating EVM chain definitions specifically.
 * This schema extends the base schema with EVM-specific properties.
 *
 * @example
 * ```typescript
 * import { evmChainDefinitionSchema } from '@core/chains/validation'
 * import { Blockchain } from '@core/chains'
 *
 * const ethereumChain = {
 *   type: 'evm',
 *   chain: Blockchain.Ethereum,
 *   name: 'Ethereum',
 *   chainId: 1,
 *   nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
 *   isTestnet: false,
 *   explorerUrl: 'https://etherscan.io/tx/{hash}',
 *   usdcAddress: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
 *   eurcAddress: null,
 *   cctp: null
 * }
 *
 * const result = evmChainDefinitionSchema.safeParse(ethereumChain)
 * if (result.success) {
 *   console.log('EVM chain definition is valid')
 * } else {
 *   console.error('Validation failed:', result.error)
 * }
 * ```
 */ const evmChainDefinitionSchema = baseChainDefinitionSchema.extend({
    type: zod.z.literal('evm'),
    chainId: zod.z.number({
        required_error: 'EVM chains must have a chainId. Please provide a valid EVM chain ID.',
        invalid_type_error: 'EVM chain ID must be a number.'
    })
}).strict() //// Reject any additional properties not defined in the schema
;
/**
 * Zod schema for validating non-EVM chain definitions.
 * This schema extends the base schema with non-EVM specific properties.
 */ const nonEvmChainDefinitionSchema = baseChainDefinitionSchema.extend({
    type: zod.z.enum([
        'algorand',
        'avalanche',
        'solana',
        'aptos',
        'near',
        'stellar',
        'sui',
        'hedera',
        'noble',
        'polkadot'
    ])
}).strict() // Reject any additional properties not defined in the schema
;
/**
 * Discriminated union schema for all chain definitions.
 * This schema validates different chain types based on their 'type' field.
 *
 * @example
 * ```typescript
 * import { chainDefinitionSchema } from '@core/chains/validation'
 * import { Blockchain } from '@core/chains'
 *
 * // EVM chain
 * chainDefinitionSchema.parse({
 *   type: 'evm',
 *   chain: Blockchain.Ethereum,
 *   chainId: 1,
 *   // ... other properties
 * })
 *
 * // Non-EVM chain
 * chainDefinitionSchema.parse({
 *   type: 'solana',
 *   chain: Blockchain.Solana,
 *   // ... other properties (no chainId)
 * })
 * ```
 */ const chainDefinitionSchema = zod.z.discriminatedUnion('type', [
    evmChainDefinitionSchema,
    nonEvmChainDefinitionSchema
]);
/**
 * Zod schema for validating chain identifiers.
 * This schema accepts either a string blockchain identifier, a Blockchain enum value,
 * or a full ChainDefinition object.
 *
 * @example
 * ```typescript
 * import { chainIdentifierSchema } from '@core/chains/validation'
 * import { Blockchain, Ethereum } from '@core/chains'
 *
 * // All of these are valid:
 * chainIdentifierSchema.parse('Ethereum')
 * chainIdentifierSchema.parse(Blockchain.Ethereum)
 * chainIdentifierSchema.parse(Ethereum)
 * ```
 */ zod.z.union([
    zod.z.string().refine((val)=>val in Blockchain, 'Must be a valid Blockchain enum value as string'),
    zod.z.nativeEnum(Blockchain),
    chainDefinitionSchema
]);
/**
 * Zod schema for validating swap-specific chain identifiers.
 *
 * Validates chains based on:
 * - CCTPv2 support (adapter contract deployed)
 * - Mainnet only (no testnets)
 * - At least one supported token available
 *
 */ zod.z.union([
    // String blockchain identifier (accepts SwapChain enum values)
    zod.z.string().refine((val)=>val in SwapChain, (val)=>({
            message: `"${val}" is not a supported swap chain. ` + `Supported chains: ${Object.values(SwapChain).join(', ')}`
        })),
    // SwapChain enum
    zod.z.nativeEnum(SwapChain),
    // ChainDefinition object (checks if chain.chain is in SwapChain)
    chainDefinitionSchema.refine((chain)=>chain.chain in SwapChain, (chain)=>({
            message: `"${chain.chain}" is not a supported swap chain. ` + `Supported chains: ${Object.values(SwapChain).join(', ')}`
        }))
]);
/**
 * Zod schema for validating bridge chain identifiers.
 *
 * This schema validates that the provided chain is supported for CCTPv2 bridging.
 * It accepts either a BridgeChain enum value, a string matching a BridgeChain value,
 * or a ChainDefinition for a supported chain.
 *
 * Use this schema when validating chain parameters for bridge operations to ensure
 * only CCTPv2-supported chains are accepted at runtime.
 *
 * @example
 * ```typescript
 * import { bridgeChainIdentifierSchema } from '@core/chains/validation'
 * import { BridgeChain, Chains } from '@core/chains'
 *
 * // Valid - BridgeChain enum value
 * bridgeChainIdentifierSchema.parse(BridgeChain.Ethereum)
 *
 * // Valid - string literal
 * bridgeChainIdentifierSchema.parse('Base_Sepolia')
 *
 * // Valid - ChainDefinition (validated by CCTP support)
 * bridgeChainIdentifierSchema.parse(Chains.Solana)
 *
 * // Invalid - Algorand is not in BridgeChain (throws ZodError)
 * bridgeChainIdentifierSchema.parse('Algorand')
 * ```
 *
 * @see {@link BridgeChain} for the enum of supported chains.
 */ zod.z.union([
    zod.z.string().refine((val)=>val in exports.BridgeChain, (val)=>({
            message: `Chain "${val}" is not supported for bridging. Only chains in the BridgeChain enum support CCTPv2 bridging.`
        })),
    chainDefinitionSchema.refine((chainDef)=>chainDef.chain in exports.BridgeChain, (chainDef)=>({
            message: `Chain "${chainDef.name}" (${chainDef.chain}) is not supported for bridging. Only chains in the BridgeChain enum support CCTPv2 bridging.`
        }))
]);
/**
 * Zod schema for validating earn-specific chain identifiers.
 *
 * Validate that the provided chain is supported for earn (vault
 * deposit/withdraw) operations. Currently only Arc Testnet is
 * supported.
 *
 * Accept an EarnChain enum value, a matching string literal, or
 * a ChainDefinition for a supported chain.
 *
 * @example
 * ```typescript
 * import { earnChainIdentifierSchema } from '@core/chains'
 * import { ArcTestnet, EarnChain } from '@core/chains'
 *
 * // Valid
 * earnChainIdentifierSchema.parse(EarnChain.Arc_Testnet)
 * earnChainIdentifierSchema.parse('Arc_Testnet')
 * earnChainIdentifierSchema.parse(ArcTestnet)
 *
 * // Invalid (throws ZodError)
 * earnChainIdentifierSchema.parse('Solana')
 * ```
 */ zod.z.union([
    zod.z.string().refine((val)=>val in EarnChain, (val)=>({
            message: `"${val}" is not a supported earn chain. ` + `Supported chains: ${Object.values(EarnChain).join(', ')}`
        })),
    zod.z.nativeEnum(EarnChain),
    chainDefinitionSchema.refine((chain)=>chain.chain in EarnChain, (chain)=>({
            message: `"${chain.chain}" is not a supported earn chain. ` + `Supported chains: ${Object.values(EarnChain).join(', ')}`
        }))
]);
// Widened views so `.includes()` accepts arbitrary strings under strict TS.
const EARN_BRIDGE_SOURCE_CHAIN_VALUES = EARN_BRIDGE_SOURCE_BLOCKCHAINS;
const EARN_BRIDGE_DESTINATION_CHAIN_VALUES = EARN_BRIDGE_DESTINATION_BLOCKCHAINS;
/**
 * Zod schema for validating the source chain of a cross-chain Earn deposit.
 *
 * Accept a supported source Blockchain value, a matching string literal, or a
 * ChainDefinition for a supported source chain. Source chains are Ethereum
 * Sepolia, Arbitrum Sepolia, and Base Sepolia.
 *
 * @example
 * ```typescript
 * import { earnBridgeSourceChainIdentifierSchema } from '@core/chains'
 *
 * // Valid
 * earnBridgeSourceChainIdentifierSchema.parse('Ethereum_Sepolia')
 *
 * // Invalid (throws ZodError)
 * earnBridgeSourceChainIdentifierSchema.parse('Arc_Testnet')
 * ```
 */ zod.z.union([
    zod.z.string().refine((val)=>EARN_BRIDGE_SOURCE_CHAIN_VALUES.includes(val), (val)=>({
            message: `"${val}" is not a supported cross-chain Earn source chain. ` + `Supported chains: ${EARN_BRIDGE_SOURCE_BLOCKCHAINS.join(', ')}`
        })),
    chainDefinitionSchema.refine((chain)=>EARN_BRIDGE_SOURCE_CHAIN_VALUES.includes(chain.chain), (chain)=>({
            message: `"${chain.chain}" is not a supported cross-chain Earn source chain. ` + `Supported chains: ${EARN_BRIDGE_SOURCE_BLOCKCHAINS.join(', ')}`
        }))
]);
/**
 * Zod schema for validating the destination chain of a cross-chain Earn
 * deposit.
 *
 * Accept a supported destination Blockchain value, a matching string literal,
 * or a ChainDefinition for a supported destination chain. Currently only Arc
 * Testnet is supported.
 *
 * @example
 * ```typescript
 * import { earnBridgeDestinationChainIdentifierSchema } from '@core/chains'
 *
 * // Valid
 * earnBridgeDestinationChainIdentifierSchema.parse('Arc_Testnet')
 *
 * // Invalid (throws ZodError)
 * earnBridgeDestinationChainIdentifierSchema.parse('Ethereum_Sepolia')
 * ```
 */ zod.z.union([
    zod.z.string().refine((val)=>EARN_BRIDGE_DESTINATION_CHAIN_VALUES.includes(val), (val)=>({
            message: `"${val}" is not a supported cross-chain Earn destination chain. ` + `Supported chains: ${EARN_BRIDGE_DESTINATION_BLOCKCHAINS.join(', ')}`
        })),
    chainDefinitionSchema.refine((chain)=>EARN_BRIDGE_DESTINATION_CHAIN_VALUES.includes(chain.chain), (chain)=>({
            message: `"${chain.chain}" is not a supported cross-chain Earn destination chain. ` + `Supported chains: ${EARN_BRIDGE_DESTINATION_BLOCKCHAINS.join(', ')}`
        }))
]);
/**
 * Zod schema for validating unified balance chain identifiers.
 *
 * This schema validates that the provided chain is supported for unified balance operations.
 * It accepts either a UnifiedBalanceChain enum value, a string matching a UnifiedBalanceChain value,
 * or a ChainDefinition for a supported chain.
 *
 * Use this schema when validating chain parameters for unified balance operations to ensure
 * only Gateway V1-supported chains are accepted at runtime.
 *
 * @example
 * ```typescript
 * import { unifiedBalanceChainIdentifierSchema } from '@core/chains/validation'
 * import { UnifiedBalanceChain, Chains } from '@core/chains'
 *
 * // Valid - UnifiedBalanceChain enum value
 * unifiedBalanceChainIdentifierSchema.parse(UnifiedBalanceChain.Ethereum)
 *
 * // Valid - string literal
 * unifiedBalanceChainIdentifierSchema.parse('Ethereum')
 *
 * // Invalid - Algorand is not in UnifiedBalanceChain (throws ZodError)
 * unifiedBalanceChainIdentifierSchema.parse('Algorand')
 * ```
 *
 * @see {@link UnifiedBalanceChain} for the enum of supported chains.
 */ const supportedUnifiedBalanceChains = Object.keys(UnifiedBalanceChain).join(', ');
zod.z.union([
    zod.z.string().refine((val)=>val in UnifiedBalanceChain, (val)=>({
            message: `Chain "${val}" is not supported for unified balance operations. Supported chains: ${supportedUnifiedBalanceChains}.`
        })),
    chainDefinitionSchema.refine((chainDef)=>chainDef.chain in UnifiedBalanceChain, (chainDef)=>({
            message: `Chain "${chainDef.name}" (${chainDef.chain}) is not supported for unified balance operations. Supported chains: ${supportedUnifiedBalanceChains}.`
        }))
]);

// Internal enum used after input normalization.
const swapTokenEnumSchema = zod.z.enum([
    ...Object.keys(SWAP_TOKEN_REGISTRY),
    NATIVE_TOKEN
]);
/**
 * Zod schema for validating supported swap token symbols.
 *
 * Accepts any token symbol from the SWAP_TOKEN_REGISTRY plus NATIVE.
 * Input matching is case-insensitive and normalized to uppercase.
 *
 * @example
 * ```typescript
 * import { supportedSwapTokenSchema } from '@core/chains'
 *
 * const result = supportedSwapTokenSchema.safeParse('USDC')
 * if (result.success) {
 *   console.log('Valid swap token:', result.data)
 * }
 * ```
 */ zod.z.string().transform((value)=>value.toUpperCase()).pipe(swapTokenEnumSchema);

exports.Arbitrum = Arbitrum;
exports.ArbitrumSepolia = ArbitrumSepolia;
exports.ArcTestnet = ArcTestnet;
exports.Avalanche = Avalanche;
exports.AvalancheFuji = AvalancheFuji;
exports.Base = Base;
exports.BaseSepolia = BaseSepolia;
exports.Codex = Codex;
exports.CodexTestnet = CodexTestnet;
exports.Cronos = Cronos;
exports.CronosTestnet = CronosTestnet;
exports.Edge = Edge;
exports.EdgeTestnet = EdgeTestnet;
exports.Ethereum = Ethereum;
exports.EthereumSepolia = EthereumSepolia;
exports.HyperEVM = HyperEVM;
exports.HyperEVMTestnet = HyperEVMTestnet;
exports.Injective = Injective;
exports.InjectiveTestnet = InjectiveTestnet;
exports.Ink = Ink;
exports.InkTestnet = InkTestnet;
exports.Linea = Linea;
exports.LineaSepolia = LineaSepolia;
exports.Monad = Monad;
exports.MonadTestnet = MonadTestnet;
exports.Morph = Morph;
exports.MorphTestnet = MorphTestnet;
exports.Optimism = Optimism;
exports.OptimismSepolia = OptimismSepolia;
exports.Pharos = Pharos;
exports.PharosTestnet = PharosTestnet;
exports.Plume = Plume;
exports.PlumeTestnet = PlumeTestnet;
exports.Polygon = Polygon;
exports.PolygonAmoy = PolygonAmoy;
exports.Sei = Sei;
exports.SeiTestnet = SeiTestnet;
exports.Solana = Solana;
exports.SolanaDevnet = SolanaDevnet;
exports.Sonic = Sonic;
exports.SonicTestnet = SonicTestnet;
exports.Unichain = Unichain;
exports.UnichainSepolia = UnichainSepolia;
exports.WorldChain = WorldChain;
exports.WorldChainSepolia = WorldChainSepolia;
exports.XDC = XDC;
exports.XDCApothem = XDCApothem;
